Optimise the manifest update

jgraham opened the pull request:
Not ready to merge etc. Mostly concerned with making no-op updates as fast as possible.

Cactusmachete committed:
Use testfile mtimes to pre-filter files iterated over for the manifest update

jgraham committed:
WIP

jgraham committed:
Update the .gitignore file

jgraham committed:
Allow the gitignore filter to work on name components only.

We end up with a lot of rules like (?:.*)/.*\.ext which are basically
trying to find the last component in a path and match against
that. These are rather slow to run so the easiest thing tdo is just
pass in the last component of the path when we know that's the only
thing the rule can match.

The changes to surrounding code to use this API will be made in future
commits.

jgraham committed:
Update the gitignore implementation to work as an iterator filter.

This updates the gitignore implemenation to take input like os.walk
but with additional stat data for the files. It also makes several
useful optimistaions:

 * Avoid using regex when just matching a literal
 * Identify patterns that can only match the final component of a path
   and run those against that component rather than the full path.
 * Add the possibility of providing a dictionary of paths to gitignore
   statuses as a cache.

This dramatically reduces the amount of time we spend in gitignore
processing when updating the manifest.

jgraham committed:
Update the lint to the new gitignore API

jgraham committed:
Add manifest caches for the mtime and gitignore rules.

When processing the manifest using the worktree, instead of reading
all files to see if the content changed, instead only process files where
the mtime has been updated since the previous run. Also cache the
result of running gitignore, so we can save a couple of seconds
processing the gitignore rules.

jgraham committed:
Add a custom implementation of os.walk

Compared to the normal os.walk this has a couple of differences:

* It returns lists of (name, stat) for filenames and directories,
  allowing callers to reuse the stat data without going back to the
  system to re-request it.

* Directories are always returned as paths relative to the root, and
  the root itself is returned as the empty string.

* It is non-recursive.

There are also a few features missing that aren't required for our use
cases.

gsnedders reviewed (commented):
I tend to be a bit weary of fixing performance just by writing a bunch of caches versus fixing the underlying performance issues, which is essentially what this does for `gitignore`.

On the mtime side, I wonder if we should be using the user's cache directory ($XDG_CACHE_HOME or whatever), rather than introducing our own cache files in random places in the filesystem?

jgraham commented on tools/manifest/commands.json:
```diff
@@ -1,6 +1,5 @@
 {"manifest":
  {"path": "update.py", "script": "run", "parser": "create_parser", "help": "Update the MANIFEST.json file",
-  "virtualenv": false},
+  "virtualenv": false, "install": ["ujson"]},
```
I didn't mean to add this. I intend to use ujson if it's available but not to require it. For a no-op update the time saving is almost exactly cancelled by the time to setup the virtualenv.

jgraham commented on tools/manifest/vcs.py:
```diff
@@ -70,27 +75,168 @@ def __iter__(self):
+            self.data[key] = value
+
+
+def walk(root):
```
Just because it's one less required dependency, and a C one at that. Also since we typically require the stat data anyway it's not clear how much of a win it will be.

jgraham committed:
Fix the cache lifecycle

The caches weren't being invalidated when the manifest itself
changed. To fix this the manifest itself has to be written before the
cache and the cache has to include data about the manifest that it's
associated with (the mtime and path are used for this purpose).

To make all this work requires a single method that can load the
manifest, update it, write the manifest and write the
caches. Therefore we introduce a single load_and_update method that is
intended to replace all previous use of the load() or update() methods
(and as a bonus handles manifest version mismatches in a single
place).

jgraham committed:
Fix the .gitignore rules

jgraham committed:
Pass the manifest file directly into wpttest.from_manifest

jgraham committed:
Update manifest before tests

jgraham commented:
I think this is bascially ready for review. The branch is a bit of a mess and the cache stuff probably needs better tests, but for my sanity I think I need to work on something else for a bit :)

Ms2ger commented on tools/gitignore/gitignore.py:
```diff
@@ -1,21 +1,27 @@
     else:
-        any_char = "."
+        # By default match the entire path up to a /
+        # but if / doesn't appear in the pattern we will mark is as
```
We will what now?

Ms2ger commented on tools/wpt/wpt.py:
```diff
@@ -133,11 +133,7 @@ def main(prog=None, argv=None):
-                pdb.post_mortem()
-            else:
-                raise
+            raise
```
Do you still need the try/except?

jgraham commented:
Since there's basically no traction on getting this landed here, I created gecko bug https://bugzilla.mozilla.org/show_bug.cgi?id=1497898 with these patches applied to mozilla-central. I really don't mind which way we land this, but it is important to me that we either land this or something functionally equivalent (i.e. something that can drop the zero-change manifest update time using the worktree to around 1s).

foolip commented:
Please ignore the CircleCI failure on this PR. I was experimenting with CircleCI on a branch and did not expect it to show up on all PRs. I've revoked access for CircleCI, and it should be possible to merge even with this spurious failure. Please let me know if there are problems.

jgraham commented:
Landed via #13618
